crates/ra_assists/src/handlers/add_function.rs

```diff
@@ -0,0 +1,716 @@
+            add_function,
+            r"
+struct Baz;
+impl Baz {
+    fn foo(&self) -> Baz {
+        ba<|>r(self.baz())
+    }
+    fn baz(&self) -> Baz {
+        Baz
+    }
+}
+",
+            r"
+struct Baz;
+impl Baz {
+    fn foo(&self) -> Baz {
+        bar(self.baz())
+    }
+    fn baz(&self) -> Baz {
+        Baz
+    }
+}
+
+fn bar(baz: Baz) {
+    <|>todo!()
+}
+",
+        )
+    }
+
+    #[test]
+    fn add_function_with_string_literal_arg() {
+        check_assist(
+            add_function,
+            r#"
+fn foo() {
+    <|>bar("bar")
+}
+"#,
+            r#"
+fn foo() {
+    bar("bar")
+}
+
+fn bar(arg: &str) {
+    <|>todo!()
+}
+"#,
+        )
+    }
+
+    #[test]
+    fn add_function_with_char_literal_arg() {
+        check_assist(
+            add_function,
+            r#"
+fn foo() {
+    <|>bar('x')
+}
+"#,
+            r#"
+fn foo() {
+    bar('x')
+}
+
+fn bar(arg: char) {
+    <|>todo!()
+}
+"#,
+        )
+    }
+
+    #[test]
+    fn add_function_with_int_literal_arg() {
+        check_assist(
+            add_function,
+            r"
+fn foo() {
+    <|>bar(42)
+}
+",
+            r"
+fn foo() {
+    bar(42)
+}
+
+fn bar(arg: i32) {
+    <|>todo!()
+}
+",
+        )
+    }
+
+    #[test]
+    fn add_function_with_cast_int_literal_arg() {
+        check_assist(
+            add_function,
+            r"
+fn foo() {
+    <|>bar(42 as u8)
+}
+",
+            r"
+fn foo() {
+    bar(42 as u8)
+}
+
+fn bar(arg: u8) {
+    <|>todo!()
+}
+",
+        )
+    }
+
+    #[test]
+    fn name_of_cast_variable_is_used() {
+        // Ensures that the name of the cast type isn't used
+        // in the generated function signature.
+        check_assist(
+            add_function,
+            r"
+fn foo() {
+    let x = 42;
+    bar<|>(x as u8)
+}
+",
+            r"
+fn foo() {
+    let x = 42;
+    bar(x as u8)
+}
+
+fn bar(x: u8) {
+    <|>todo!()
+}
+",
+        )
+    }
+
+    #[test]
+    fn add_function_with_variable_arg() {
+        check_assist(
+            add_function,
+            r"
+fn foo() {
+    let worble = ();
+    <|>bar(worble)
+}
+",
+            r"
+fn foo() {
+    let worble = ();
+    bar(worble)
+}
+
+fn bar(worble: ()) {
+    <|>todo!()
+}
+",
+        )
+    }
+
+    #[test]
+    fn add_function_with_impl_trait_arg() {
+        check_assist(
+            add_function,
+            r"
+trait Foo {}
+fn foo() -> impl Foo {
+    todo!()
+}
+fn baz() {
+    <|>bar(foo())
+}
+",
+            r"
+trait Foo {}
+fn foo() -> impl Foo {
+    todo!()
+}
+fn baz() {
+    bar(foo())
+}
+
+fn bar(foo: impl Foo) {
+    <|>todo!()
+}
+",
+        )
+    }
+
+    #[test]
+    fn add_function_with_fn_arg() {
+        check_assist(
+            add_function,
+            r"
+struct Baz;
+impl Baz {
+    fn new() -> Self { Baz }
+}
+fn foo() {
+    <|>bar(Baz::new);
+}
+",
+            r"
+struct Baz;
+impl Baz {
+    fn new() -> Self { Baz }
+}
+fn foo() {
+    bar(Baz::new);
+}
+
+fn bar(arg: fn() -> Baz) {
```

Comment:
This fails because the type of `Baz::new` is printed as `fn new() -> Baz`.
Any idea how I can print the type without a name?

Reply:
Reading the implementation of `HirDisplay` for `ApplicationTy`, it looks like that code is written for error messages and not for generating valid Rust code. Maybe this is the point where I start migrating to the `ast::make` API?

Reply:
Aw, the "Insert explicit type" assist has the same issue :/ that kinda makes me think we can leave that case unhandled at first

Reply:
Yeah, I would at least punt that to another PR. We would also want to qualify (or import) names if necessary. I'm not sure whether we should build all this into the `HirDisplay` infrastructure, or rather introduce a new "AST printing" functionality :thinking:

Reply:
It looks like  `TreeDiff::into_text_edit` (which is used by some other assists) uses the `Display` implementation of `rowan::SyntaxElement<RustLanguage>`... I guess that doesn't help much with types, huh.

Reply:
Yeah, going from our syntax tree to a string is trivial, the problem is going from higher-level representations to syntax :)

Reply:
i'm starting to learn that as well :sweat_smile: